File: python/crosssection.py
import math


class CrossSection(object):
    
    def __init__(self, nodes):
        self.__nodes = self.__connect(nodes)
        self.__section_area = 0.0  # F
        self.__sectorial_static_moment = 0.0  # Sw
        self.__sectorial_inertia_moment = 0.0  # Iw
        self.__sectorial_linear_static_moment = None  # Swx,y (Point)
        self.__inertia_moment = None  # Ix,y (Point)
        self.__gravity_center = None  # Point
        self.__rigidity_center = None  # Point
        self.__pole = None  # Point

    def get_section_area(self, lazy=True):
        """Return a total area of the cross section.
        """
        if self.__section_area > 0 and lazy:
            return self.__section_area

        self.__section_area = 0.0
        self.__traverse_nodes(self.__nodes[0], self.__section_area_callback)

        return self.__section_area

    def __section_area_callback(self, node):
        if node.parent is None: return

        ds = node.distance_to(node.parent)
        thickness = 0.5 * (node.thickness + node.parent.thickness)
        self.__section_area += thickness * ds

    def get_gravity_center(self, lazy=True):
        """Return a point where the center of gravity is located.
        """
        if self.__gravity_center and lazy:
            return self.__gravity_center

        self.__gravity_center = Point()
        self.__traverse_nodes(self.__nodes[0], self.__gravity_center_callback)

        self.__gravity_center.x /= self.get_section_area()
        self.__gravity_center.y /= self.get_section_area()

        return self.__gravity_center

    def __gravity_center_callback(self, node):
        if node.parent is None: return

        ds = node.distance_to(node.parent)
        thickness = 0.5 * (node.thickness + node.parent.thickness)

        self.__gravity_center.x += 0.5 * (
            node.x + node.parent.x
        ) * thickness * ds

        self.__gravity_center.y += 0.5 * (
            node.y + node.parent.y
        ) * thickness * ds

    def get_inertia_moment(self, lazy=True):
        """Return a point whose coordinates represent main moments of inertia
        calculated for the corresponding axis (Ix,y).
        """
        if self.__inertia_moment and lazy:
            return self.__inertia_moment

        self.__inertia_moment = Point()
        self.__traverse_nodes(self.__nodes[0], self.__inertia_moment_callback)

        sa = self.get_section_area()
        gc = self.get_gravity_center()

        self.__inertia_moment.x -= sa * gc.y**2;
        self.__inertia_moment.y -= sa * gc.x**2;

        return self.__inertia_moment

    def __inertia_moment_callback(self, node):
        if node.parent is None: return

        ds = node.distance_to(node.parent)
        thickness = 0.5 * (node.thickness + node.parent.thickness)

        self.__inertia_moment.x += 0.5 * (
            node.y**2 + node.parent.y**2
        ) * thickness * ds

        self.__inertia_moment.y += 0.5 * (
            node.x**2 + node.parent.x**2
        ) * thickness * ds

    def get_polar_inertia_moment(self, lazy=True):
        """Return a polar moment of inertia (Ip).
        """
        return (self.get_inertia_moment(lazy).x
                + self.get_inertia_moment(lazy).y)

    def __get_area_sign(self, start_point, end_point):
        """Return a sign of the sectorial area defined by given two points and
        a reference origin.
        """
        sign = lambda x: (x > 0) - (x < 0)
        return sign(
            self.__get_angle(end_point) - self.__get_angle(start_point)
        )

    def __get_angle(self, point):
        """Return an angle between X-axis and a radius vector defined by a
        given point and the reference origin.
        """
        angle = math.degrees(math.atan2(point.y, point.x))

        if round(angle, 2) < 0:
            angle += 360

        return angle

    def __get_triangle_area(self, a, b, c):
        """Return an area of a triangle formed by given three points using the
        Heron's formula.
        """
        ab = a.distance_to(b)
        bc = b.distance_to(c)
        ca = c.distance_to(a)
        p = (ab + bc + ca) / 2
        return (p*(p - ab)*(p - bc)*(p - ca))**0.5

    def __update_sectorial_area(self, root_node, pole):
        """Update values of the sectorial area in nodes given a root node to
        start from and a pole point.
        """
        for node in self.__nodes:
            node.sectorial_area = 0.0
        self.__pole = pole
        self.__traverse_nodes(root_node, self.__update_sectorial_area_callback)

    def __update_sectorial_area_callback(self, node):
        if node.parent is None: return

        node_pos = Point(
            node.x - self.__pole.x,
            node.y - self.__pole.y
        )

        node_parent_pos = Point(
            node.parent.x - self.__pole.x,
            node.parent.y - self.__pole.y
        )

        origin_pos = Point()

        area_sign = self.__get_area_sign(node_parent_pos, node_pos)
        area = self.__get_triangle_area(node_parent_pos, node_pos, origin_pos)
        area_inc = area_sign * area * 2

        node.sectorial_area = node.parent.sectorial_area + area_inc

    def get_sectorial_static_moment(self, root_node, pole):
        """Return a sectorial static moment of inertia (Sw).
        """
        self.__sectorial_static_moment = 0.0
        self.__update_sectorial_area(root_node, pole)
        self.__traverse_nodes(root_node, self.__sectorial_static_moment_callback)
        return self.__sectorial_static_moment

    def __sectorial_static_moment_callback(self, node):
        if node.parent is None: return

        node_pos = Point(
            node.x - self.__pole.x,
            node.y - self.__pole.y
        )

        node_parent_pos = Point(
            node.parent.x - self.__pole.x,
            node.parent.y - self.__pole.y
        )

        ds = node_pos.distance_to(node_parent_pos)
        thickness = 0.5 * (node.thickness + node.parent.thickness)

        self.__sectorial_static_moment += 0.5 * (
            node.sectorial_area + node.parent.sectorial_area
        ) * thickness * ds

    def get_sectorial_linear_static_moment(self, root_node, pole):
        """Return a point whose coordinates represent sectorial linear static
        moments of inertia calculated for the corresponding axes (Swx,y).
        """
        self.__sectorial_linear_static_moment = Point()
        self.__update_sectorial_area(root_node, pole)
        self.__traverse_nodes(root_node, self.__sectorial_linear_static_moment_callback)
        return self.__sectorial_linear_static_moment

    def __sectorial_linear_static_moment_callback(self, node):
        if node.parent is None: return

        node_pos = Point(
            node.x - self.get_gravity_center().x,
            node.y - self.get_gravity_center().y
        )

        node_parent_pos = Point(
            node.parent.x - self.get_gravity_center().x,
            node.parent.y - self.get_gravity_center().y
        )

        ds = node_pos.distance_to(node_parent_pos)
        thickness = 0.5 * (node.thickness + node.parent.thickness)

        self.__sectorial_linear_static_moment.x += 0.5 * (
            node_pos.y * node.sectorial_area +
            node_parent_pos.y * node.parent.sectorial_area
        ) * thickness * ds

        self.__sectorial_linear_static_moment.y += 0.5 * (
            node_pos.x * node.sectorial_area +
            node_parent_pos.x * node.parent.sectorial_area
        ) * thickness * ds

    def get_rigidity_center(self, lazy=True):
        """Return a point where the rigidity center is located.
        """
        if self.__rigidity_center and lazy:
            return self.__rigidity_center

        pole = Point()

        inertia_moment = self.get_inertia_moment()
        sectorial_linear_static_moment = \
            self.get_sectorial_linear_static_moment(self.__nodes[0], pole)

        self.__rigidity_center = Point(
            pole.x + sectorial_linear_static_moment.x / inertia_moment.x,
            pole.y - sectorial_linear_static_moment.y / inertia_moment.y
        )

        return self.__rigidity_center

    def get_sectorial_inertia_moment(self, lazy=True):
        """Return a sectorial moment of inertia (Iw).
        """
        if self.__sectorial_inertia_moment and lazy:
            return self.__sectorial_inertia_moment

        root_node = None
        pole = self.get_rigidity_center()

        s_min = float('inf')
        for node in self.__nodes:
            ssm = self.get_sectorial_static_moment(node, pole)
            slsm = self.get_sectorial_linear_static_moment(node, pole)
            s = abs(ssm) + abs(slsm.x) + abs(slsm.y)
            if s < s_min:
                s_min = s
                root_node = node

        self.__update_sectorial_area(root_node, pole)

        self.__sectorial_inertia_moment = 0.0
        self.__traverse_nodes(root_node, self.__sectorial_inertia_moment_callback)

        return self.__sectorial_inertia_moment

    def __sectorial_inertia_moment_callback(self, node):
        if node.parent is None: return

        node_pos = Point(
            node.x - self.get_rigidity_center().x,
            node.y - self.get_rigidity_center().y
        )

        node_parent_pos = Point(
            node.parent.x - self.get_rigidity_center().x,
            node.parent.y - self.get_rigidity_center().y
        )

        ds = node_pos.distance_to(node_parent_pos)
        thickness = 0.5 * (node.thickness + node.parent.thickness)

        self.__sectorial_inertia_moment += 0.5 * (
            node.sectorial_area**2 + node.parent.sectorial_area**2        
        ) * thickness * ds

    def __traverse_nodes(self, root_node, callback):
        """Traverse nodes using DFS invoking at each step a callback function
        which receives the current node as its parameter.
        """
        for node in self.__nodes:
            node.parent = None

        visited_nodes = []
        nodes_to_visit = [root_node]

        while nodes_to_visit:
            cur_node = nodes_to_visit.pop()
            for node in cur_node.links:
                if node not in visited_nodes:
                    node.parent = cur_node
                    nodes_to_visit.append(node)
            callback(cur_node)
            visited_nodes.append(cur_node)

    def __connect(self, nodes):
        """Connect nodes in a given list. That is, build a minimum spanning
        tree (MST) using a greedy algorithm.
        """
        disconnected_nodes = nodes[:]
        connected_nodes = [disconnected_nodes.pop()]

        while disconnected_nodes:
            node_pair = {
                'dist': float('inf'),
                'd_node': None,
                'c_node': None
            }
            for c_node in connected_nodes:
                for d_node in disconnected_nodes:
                    dist = d_node.distance_to(c_node)
                    if dist < node_pair['dist']:
                        node_pair['dist'] = dist
                        node_pair['d_node'] = d_node
                        node_pair['c_node'] = c_node

            node_pair['c_node'].connect(node_pair['d_node'])
            node_pair['d_node'].connect(node_pair['c_node'])

            connected_nodes.append(node_pair['d_node'])
            disconnected_nodes.remove(node_pair['d_node'])

        return connected_nodes


class Point(object):

    def __init__(self, x=0.0, y=0.0):
        self.x = x
        self.y = y

    def distance_to(self, other):
        return ((other.x - self.x)**2 + (other.y - self.y)**2)**0.5

    def __str__(self):
        return '({x:.2f}, {y:.2f})'.format(x=self.x, y=self.y)

    def __repr__(self):
        return '{cls}(x={x:.2f}, y={y:.2f})'.format(
            cls=self.__class__.__name__, x=self.x, y=self.y
        )


class Node(Point):

    def __init__(self, x=0.0, y=0.0, thickness=1.0):
        super(Node, self).__init__(x, y)
        self.thickness = thickness
        self.sectorial_area = 0.0
        self.links = []
        self.parent = None

    def connect(self, node):
        if node not in self.links:
            self.links.append(node)

    def disconnect(self, node):
        if node in self.links:
            self.links.remove(node)

    def __str__(self):
        return 'node: {p} linked with: {ps}'.format(
            p=super(Node, self).__str__(),
            ps=', '.join([super(Node, node).__str__() for node in self.links])
        )

    def __repr__(self):
        return '{cls}(x={x:.2f}, y={y:.2f}, thickness={thickness})'.format(
            cls=self.__class__.__name__, x=self.x, y=self.y,
            thickness=self.thickness
        )

File: java/CrossSection.java
import java.util.Stack;
import java.util.Arrays;
import java.util.ArrayList;


public class CrossSection {
    private ArrayList<Node> nodes = null;
    private double sectionArea = 0.0;
    private double sectorialStaticMoment = 0.0;
    private double sectorialInertiaMoment = 0.0;
    private Point sectorialLinearStaticMoment = null;
    private Point inertiaMoment = null;
    private Point gravityCenter = null;
    private Point rigidityCenter = null;
    private Point pole = null;

    public CrossSection(Node[] nodes) {
        this.nodes = connect(nodes);
    }

    /**
     * Return a total area of the cross section.
     *
     */
    public double getSectionArea(boolean lazy) {
        if (sectionArea > 0 && lazy) {
            return sectionArea;
        }

        sectionArea = 0.0;
        traverseNodes(nodes.get(0),
                      (Node node) -> sectionAreaCallback(node));

        return sectionArea;
    }

    public double getSectionArea() {
        return getSectionArea(true);
    }

    private void sectionAreaCallback(Node node) {
        if (node.parent == null) return;

        double ds = node.distanceTo(node.parent);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);
        sectionArea += thickness * ds;
    }

    /**
     * Return a point where the center of gravity is located.
     *
     */
    public Point getGravityCenter(boolean lazy) {
        if (gravityCenter != null && lazy) {
            return gravityCenter;
        }

        gravityCenter = new Point();
        traverseNodes(nodes.get(0),
                      (Node node) -> gravityCenterCallback(node));

        gravityCenter.x /= getSectionArea();
        gravityCenter.y /= getSectionArea();

        return gravityCenter;
    }

    public Point getGravityCenter() {
        return getGravityCenter(true);
    }

    private void gravityCenterCallback(Node node) {
        if (node.parent == null) return;

        double ds = node.distanceTo(node.parent);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);

        gravityCenter.x += 0.5 * (
            node.x + node.parent.x
        ) * thickness * ds;

        gravityCenter.y += 0.5 * (
            node.y + node.parent.y
        ) * thickness * ds;
    }

    /**
     * Return a point whose coordinates represent main moments of inertia
     * calculated for the corresponding axis (Ix,y).
     *
     */
    public Point getInertiaMoment(boolean lazy) {
        if (inertiaMoment != null && lazy) {
            return inertiaMoment;
        }

        inertiaMoment = new Point();
        traverseNodes(nodes.get(0),
                      (Node node) -> inertiaMomentCallback(node));

        double sa = getSectionArea();
        Point gc = getGravityCenter();

        inertiaMoment.x -= sa * Math.pow(gc.y, 2);
        inertiaMoment.y -= sa * Math.pow(gc.x, 2);

        return inertiaMoment;
    }

    public Point getInertiaMoment() {
        return getInertiaMoment(true);
    }

    private void inertiaMomentCallback(Node node) {
        if (node.parent == null) return;

        double ds = node.distanceTo(node.parent);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);

        inertiaMoment.x += 0.5 * (
            Math.pow(node.y, 2) + Math.pow(node.parent.y, 2)
        ) * thickness * ds;

        inertiaMoment.y += 0.5 * (
            Math.pow(node.x, 2) + Math.pow(node.parent.x, 2)
        ) * thickness * ds;
    }

    /**
     * Return a polar moment of inertia (Ip).
     *
     */
    public double getPolarInertiaMoment(boolean lazy) {
        return getInertiaMoment(lazy).x + getInertiaMoment(lazy).y;
    }

    public double getPolarInertiaMoment() {
        return getInertiaMoment(true).x + getInertiaMoment(true).y;
    }

    /**
     * Return a sign of the sectorial area defined by given two points and a
     * reference origin.
     *
     */
    private double getAreaSign(Point start, Point end) {
        return Math.signum(getAngle(end) - getAngle(start));
    }

    /**
     * Return an angle between X-axis and a radius vector defined by a given
     * point and the reference origin.
     *
     */
    private double getAngle(Point p) {
        double angle = Math.toDegrees(Math.atan2(p.y, p.x));
        angle = (double) Math.round(angle * 100) / 100;

        if (angle < 0.0) {
            angle += 360.0;
        }

        return angle;
    }

    /**
     * Return an area of a triangle formed by given three points using the
     * Heron's formula.
     *
     */
    private double getTriangleArea(Point a, Point b, Point c) {
        double ab = a.distanceTo(b);
        double bc = b.distanceTo(c);
        double ca = c.distanceTo(a);
        double p = (ab + bc + ca) / 2.0;
        return Math.sqrt(p*(p - ab)*(p - bc)*(p - ca));
    }

    /**
     * Update values of the sectorial area in nodes given a root node to start
     * from and a pole point.
     *
     */
    private void updateSectorialArea(Node rootNode, Point pole) {
        for (Node node : nodes) {
            node.sectorialArea = 0.0;
        }
        this.pole = pole;
        traverseNodes(rootNode,
                      (Node node) -> updateSectorialAreaCallback(node));
    }

    private void updateSectorialAreaCallback(Node node) {
        if (node.parent == null) return;

        Point nodePos = new Point(
            node.x - pole.x,
            node.y - pole.y
        );

        Point nodeParentPos = new Point(
            node.parent.x - pole.x,
            node.parent.y - pole.y
        );

        Point originPos = new Point();

        double areaSign = getAreaSign(nodeParentPos, nodePos);
        double area = getTriangleArea(nodeParentPos, nodePos, originPos);
        double areaInc = areaSign * area * 2.0;

        node.sectorialArea = node.parent.sectorialArea + areaInc;
    }

    /**
     * Return a sectorial static moment of inertia (Sw).
     *
     */
    public double getSectorialStaticMoment(Node rootNode, Point pole) {
        sectorialStaticMoment = 0.0;
        updateSectorialArea(rootNode, pole);
        traverseNodes(rootNode,
                      (Node node) -> sectorialStaticMomentCallback(node));
        return sectorialStaticMoment;
    }

    private void sectorialStaticMomentCallback(Node node) {
        if (node.parent == null) return;

        Point nodePos = new Point(
            node.x - pole.x,
            node.y - pole.y
        );

        Point nodeParentPos = new Point(
            node.parent.x - pole.x,
            node.parent.y - pole.y
        );

        double ds = nodePos.distanceTo(nodeParentPos);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);

        sectorialStaticMoment += 0.5 * (
            node.sectorialArea + node.parent.sectorialArea
        ) * thickness * ds;
    }

    /**
     * Return a point whose coordinates represent sectorial linear static
     * moments of inertia calculated for the corresponding axes (Swx,y).
     *
     */
    public Point getSectorialLinearStaticMoment(Node rootNode, Point pole) {
        sectorialLinearStaticMoment = new Point();
        updateSectorialArea(rootNode, pole);
        traverseNodes(rootNode,
                      (Node node) -> sectorialLinearStaticMomentCallback(node));
        return sectorialLinearStaticMoment;
    }

    private void sectorialLinearStaticMomentCallback(Node node) {
        if (node.parent == null) return;

        Point nodePos = new Point(
            node.x - getGravityCenter().x,
            node.y - getGravityCenter().y
        );

        Point nodeParentPos = new Point(
            node.parent.x - getGravityCenter().x,
            node.parent.y - getGravityCenter().y
        );

        double ds = nodePos.distanceTo(nodeParentPos);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);

        sectorialLinearStaticMoment.x += 0.5 * (
            nodePos.y * node.sectorialArea +
            nodeParentPos.y * node.parent.sectorialArea
        ) * thickness * ds;

        sectorialLinearStaticMoment.y += 0.5 * (
            nodePos.x * node.sectorialArea +
            nodeParentPos.x * node.parent.sectorialArea
        ) * thickness * ds;
    }

    /**
     * Return a point where the rigidity center is located.
     *
     */
    public Point getRigidityCenter(boolean lazy) {
        if (rigidityCenter != null && lazy) {
            return rigidityCenter;
        }

        Point pole = new Point();

        Point inertiaMoment = getInertiaMoment();
        Point sectorialLinearStaticMoment =
            getSectorialLinearStaticMoment(nodes.get(0), pole);

        rigidityCenter = new Point(
            pole.x + sectorialLinearStaticMoment.x / inertiaMoment.x,
            pole.y - sectorialLinearStaticMoment.y / inertiaMoment.y
        );

        return rigidityCenter;
    }

    public Point getRigidityCenter() {
        return getRigidityCenter(true);
    }

    /**
     * Return a sectorial moment of inertia (Iw).
     *
     */
    public double getSectorialInertiaMoment(boolean lazy) {
        if (sectorialInertiaMoment > 0.0 && lazy) {
            return sectorialInertiaMoment;
        }

        Node rootNode = null;
        Point pole = getRigidityCenter();

        double sMin = Double.POSITIVE_INFINITY;
        for (Node node : nodes) {
            double ssm = getSectorialStaticMoment(node, pole);
            Point slsm = getSectorialLinearStaticMoment(node, pole);
            double s = Math.abs(ssm) + Math.abs(slsm.x) + Math.abs(slsm.y);
            if (s < sMin) {
                sMin = s;
                rootNode = node;
            }
        }

        updateSectorialArea(rootNode, pole);

        sectorialInertiaMoment = 0.0;
        traverseNodes(rootNode,
                      (Node node) -> sectorialInertiaMomentCallback(node));

        return sectorialInertiaMoment;
    }

    public double getSectorialInertiaMoment() {
        return getSectorialInertiaMoment(true);
    }

    private void sectorialInertiaMomentCallback(Node node) {
        if (node.parent == null) return;

        Point nodePos = new Point(
            node.x - getRigidityCenter().x,
            node.y - getRigidityCenter().y
        );

        Point nodeParentPos = new Point(
            node.parent.x - getRigidityCenter().x,
            node.parent.y - getRigidityCenter().y
        );

        double ds = nodePos.distanceTo(nodeParentPos);
        double thickness = 0.5 * (node.thickness + node.parent.thickness);

        sectorialInertiaMoment += 0.5 * (
            Math.pow(node.sectorialArea, 2) +
            Math.pow(node.parent.sectorialArea, 2)
        ) * thickness * ds;
    }

    /**
     * Traverse nodes using DFS invoking at each step a callback function which
     * receives the current node as its parameter.
     *
     */
    private void traverseNodes(Node rootNode, Callback callback) {
        for (Node node : nodes) {
            node.parent = null;
        }

        ArrayList<Node> visitedNodes = new ArrayList<Node>();
        Stack<Node> nodesToVisit = new Stack<Node>();

        nodesToVisit.push(rootNode);

        while (!nodesToVisit.empty()) {
            Node curNode = nodesToVisit.pop();
            for (Node node : curNode.links) {
                if (!visitedNodes.contains(node)) {
                    node.parent = curNode;
                    nodesToVisit.push(node);
                }
            }
            callback.invoke(curNode);
            visitedNodes.add(curNode);
        }
    }

    /**
     * Connect nodes in a given list. That is, build a minimum spanning tree
     * (MST) using a greedy algorithm.
     *
     */
    private ArrayList<Node> connect(Node[] nodes) {
        ArrayList<Node> connectedNodes = new ArrayList<Node>();
        ArrayList<Node> disconnectedNodes = 
            new ArrayList<Node>(Arrays.asList(nodes));

        connectedNodes.add(disconnectedNodes.remove(nodes.length - 1));

        while (disconnectedNodes.size() > 0) {
            double minDist = Double.POSITIVE_INFINITY;
            Node dNode = null;
            Node cNode = null;

            for (Node c : connectedNodes) {
                for (Node d : disconnectedNodes) {
                    double dist = d.distanceTo(c);
                    if (dist < minDist) {
                        minDist = dist;
                        dNode = d;
                        cNode = c;
                    }
                }
            }

            cNode.connect(dNode);
            dNode.connect(cNode);

            connectedNodes.add(dNode);
            disconnectedNodes.remove(dNode);
        }

        return connectedNodes;
    }

    interface Callback {
        void invoke(Node node);
    }
}
